File: apps/mobile/stores/chat/run-stream-controller.ts
import {
  startNativeChatStream,
  startNativeGetStream,
} from "@/utils/native-stream";
import { getRunStreamPath } from "@console/api";
import type { AgentSessionEvent, RunPromptDto } from "@console/types";
import { isAbortError } from "./chat-stream-runner";

/** Reconnect attempts before giving up on a dropped run stream. */
const MAX_RECONNECT_ATTEMPTS = 3;
/** Exponential backoff between reconnect attempts (1s, 2s, 4s). */
const RECONNECT_BACKOFF_MS = [1_000, 2_000, 4_000];

export interface RunStreamControllerDeps {
  setSessions: (
    updater: (
      sessions: Record<string, import("@/types").ChatSessionState>,
    ) => Record<string, import("@/types").ChatSessionState>,
  ) => void;
  handleEvent: (event: AgentSessionEvent) => void;
  /** Append an error bubble to the transcript. */
  markError: (message: string) => void;
  /** Finalize the local run state (called exactly once per run). */
  finalize: (hadError: boolean) => void;
  baseUrl: () => string;
}

type StreamOpener = () => () => void;

/**
 * Owns the lifecycle of one session's run stream: the initial POST /run SSE
 * pipe plus reconnect-with-resume attempts against the re-attach endpoint.
 *
 * Guarantees:
 *  - `deps.finalize` runs at most once per controller lifetime.
 *  - A transport failure triggers up to MAX_RECONNECT_ATTEMPTS resumptions
 *    (`GET /run/stream?since=<lastSeq>`); the server's 409 ("no active run")
 *    or a terminal `done`/`aborted` frame finalizes cleanly instead of
 *    surfacing an error.
 *  - `cancel()` (user pressed stop) kills pending reconnect timers and marks
 *    everything inert so late callbacks can't resurrect the run UI.
 */
export class RunStreamController {
  private closeActive?: () => void;
  private reconnectTimer?: ReturnType<typeof setTimeout>;
  private finished = false;
  private userCancelled = false;
  private hadError = false;
  /** Set between a transport failure and the paired onEnd callback. */
  private awaitingEndAfterFailure = false;
  private attempts = 0;
  private lastSeq?: number;

  constructor(
    public readonly sessionId: string,
    private readonly deps: RunStreamControllerDeps,
  ) {}

  get isActive(): boolean {
    return !this.finished && !this.userCancelled;
  }

  /** Start a fresh agent run (POST /api/sessions/:id/run). */
  startRun(body: Partial<RunPromptDto> & Record<string, unknown>): void {
    const baseUrl = this.deps.baseUrl();
    this.open(() =>
      startNativeChatStream(
        `chat-${this.sessionId}-${Date.now()}`,
        `${baseUrl}/api/sessions/${this.sessionId}/run`,
        body,
        this.callbacks(),
      ),
    );
  }

  /**
   * Attach to the server-side active run. With `since`, buffered events newer
   * than that seq are replayed first; without it, go live immediately
   * (server replays from seq 0 of the current run's buffer when since=0).
   */
  attach(since?: number): void {
    const baseUrl = this.deps.baseUrl();
    this.open(() =>
      startNativeGetStream(
        `reattach-${this.sessionId}-${Date.now()}`,
        `${baseUrl}${getRunStreamPath(this.sessionId, since)}`,
        this.callbacks(),
      ),
    );
  }

  /** User-initiated stop: kill streams/timers and mark inert. */
  cancel(): void {
    this.userCancelled = true;
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    this.reconnectTimer = undefined;
    this.closeActive?.();
    this.closeActive = undefined;
  }

  private open(opener: StreamOpener): void {
    // One live stream at a time — replacing closes the previous one.
    this.closeActive?.();
    this.closeActive = opener();
  }

  private callbacks() {
    return {
      onEvent: (event: AgentSessionEvent, meta?: { seq?: number }) => {
        if (meta?.seq != null) this.lastSeq = meta.seq;

        if (event.type === "done" || event.type === "aborted") {
          // Terminal frames arrive only on extendedFrames attach streams.
          this.finish(false);
          return;
        }
        if (event.type === "error" && !isAbortError(event.error.message)) {
          this.hadError = true;
        }
        this.deps.handleEvent(event);
      },

      onError: (message: string, info?: { statusCode?: number }) => {
        if (this.userCancelled || this.finished) return;

        // 409 = the run settled server-side while we were disconnected.
        // Not a failure — finalize gracefully without reconnect churn.
        if (info?.statusCode === 409) {
          this.finish(this.hadError);
          return;
        }

        if (this.attempts < MAX_RECONNECT_ATTEMPTS) {
          this.attempts += 1;
          this.awaitingEndAfterFailure = true;
          const delay = RECONNECT_BACKOFF_MS[this.attempts - 1] ?? 4_000;
          this.reconnectTimer = setTimeout(() => {
            this.reconnectTimer = undefined;
            if (this.userCancelled || this.finished) return;
            this.attach(this.lastSeq);
          }, delay);
          return;
        }

        // Out of retries (or run gone): surface the error; the paired onEnd
        // finalizes with hadError=true.
        this.hadError = true;
        this.deps.markError(message);
      },

      onEnd: (abortedFlag: boolean) => {
        if (this.userCancelled || this.finished) return;
        if (this.awaitingEndAfterFailure) {
          // Native wrapper reports failure via onError then onEnd(false);
          // a reconnect is already scheduled — don't finalize here.
          this.awaitingEndAfterFailure = false;
          return;
        }
        // Normal stream end: for POST /run pipes the server closes after the
        // run completes; attach streams close once the run settles. Either
        // way, end-of-stream means the run is done locally.
        void abortedFlag;
        this.finish(this.hadError);
      },
    };
  }

  private finish(hadError: boolean): void {
    if (this.finished) return;
    this.finished = true;
    if (this.reconnectTimer) clearTimeout(this.reconnectTimer);
    this.reconnectTimer = undefined;
    this.closeActive?.();
    this.closeActive = undefined;
    this.deps.finalize(hadError);
  }
}

// --- Per-session registry ---

const controllers = new Map<string, RunStreamController>();

export function getOrCreateController(
  sessionId: string,
  deps: RunStreamControllerDeps,
): RunStreamController {
  let controller = controllers.get(sessionId);
  if (!controller) {
    controller = new RunStreamController(sessionId, deps);
    controllers.set(sessionId, controller);
  }
  return controller;
}

export function getController(sessionId: string): RunStreamController | undefined {
  return controllers.get(sessionId);
}

export function removeController(sessionId: string): void {
  controllers.delete(sessionId);
}

File: apps/android/app/src/main/kotlin/com/console/mobile/data/stream/RunStreamController.kt
package com.console.mobile.data.stream

import com.console.mobile.core.chat.isAbortError
import com.console.mobile.data.model.AgentSessionEvent
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Owns one session's run stream lifecycle: the initial POST /run pipe plus
 * reconnect-with-resume against GET /run/stream?since=<lastSeq>.
 *
 * Port of apps/mobile/stores/chat/run-stream-controller.ts RunStreamController:
 *  - finalize runs at most once per controller lifetime
 *  - transport failure -> up to 3 resumptions with 1s/2s/4s backoff
 *  - 409 (no active run) or terminal done/aborted frame -> finish cleanly
 *  - cancel() (user stop) kills timers/streams; late callbacks are inert
 */
class RunStreamController(
    val sessionId: String,
    private val deps: Deps,
) {
    interface Deps {
        fun handleEvent(event: AgentSessionEvent)
        fun markError(message: String)
        fun finalize(hadError: Boolean)
        fun baseUrl(): String
        fun authToken(): String?
        fun runBodyJson(): String
        fun scope(): kotlinx.coroutines.CoroutineScope
    }

    companion object {
        const val MAX_RECONNECT_ATTEMPTS = 3
        val RECONNECT_BACKOFF_MS = longArrayOf(1_000, 2_000, 4_000)
    }

    private var collectJob: Job? = null
    private var reconnectJob: Job? = null
    private var finished = false
    private var userCancelled = false
    private var hadError = false
    private var awaitingEndAfterFailure = false
    private var attempts = 0
    private var lastSeq: Long? = null
    private var stream: ChatStreamClient? = null

    val isActive: Boolean get() = !finished && !userCancelled
    val lastSeqValue: Long? get() = lastSeq

    fun attachStream(client: ChatStreamClient) {
        stream = client
    }

    /** Start a fresh agent run (POST /api/sessions/:id/run). */
    fun startRun(client: ChatStreamClient) {
        stream = client
        open { client.startRun(deps.baseUrl(), sessionId, deps.runBodyJson(), deps.authToken()) }
    }

    /** Attach to the server-side active run, replaying events newer than since. */
    fun attach(since: Long? = null, client: ChatStreamClient = stream ?: return) {
        stream = client
        open { client.attach(deps.baseUrl(), sessionId, since, deps.authToken()) }
    }

    /** User-initiated stop. */
    fun cancel() {
        userCancelled = true
        reconnectJob?.cancel()
        reconnectJob = null
        collectJob?.cancel()
        collectJob = null
    }

    fun onEvent(event: AgentSessionEvent, seq: Long?) {
        if (userCancelled || finished) return
        if (seq != null) lastSeq = seq
        if (event.type == "done" || event.type == "aborted") {
            finish(false)
            return
        }
        if (event.type == "error" && !isAbortError(event.error?.message.orEmpty())) {
            hadError = true
        }
        deps.handleEvent(event)
    }

    fun onError(message: String, statusCode: Int?) {
        if (userCancelled || finished) return
        // 409 = run settled server-side while disconnected -> finish gracefully.
        if (statusCode == 409) {
            finish(hadError)
            return
        }
        if (attempts < MAX_RECONNECT_ATTEMPTS) {
            attempts += 1
            awaitingEndAfterFailure = true
            val delayMs = RECONNECT_BACKOFF_MS.getOrElse(attempts - 1) { 4_000 }
            val client = stream ?: return
            reconnectJob?.cancel()
            reconnectJob = deps.scope().launch {
                delay(delayMs)
                if (userCancelled || finished) return@launch
                attach(lastSeq, client)
            }
            return
        }
        hadError = true
        deps.markError(message)
    }

    fun onEnd(aborted: Boolean) {
        if (userCancelled || finished) return
        if (awaitingEndAfterFailure) {
            // Failure already scheduled a reconnect — don't finalize here.
            awaitingEndAfterFailure = false
            return
        }
        finish(hadError)
    }

    private fun open(flows: () -> kotlinx.coroutines.flow.Flow<StreamOutcome>) {
        collectJob?.cancel()
        collectJob = deps.scope().launch {
            try {
                flows().collect { outcome ->
                    when (outcome) {
                        is StreamOutcome.Frame -> onEvent(outcome.frame.event, outcome.frame.seq)
                        is StreamOutcome.End -> when (val e = outcome.end) {
                            is StreamEnd.Aborted -> onEnd(true)
                            is StreamEnd.Completed -> onEnd(false)
                            is StreamEnd.Failed -> {
                                onError(e.message, e.statusCode)
                                // Pair with the terminal end unless a reconnect was scheduled.
                                if (!awaitingEndAfterFailure) onEnd(false)
                            }
                        }
                    }
                }
            } catch (_: kotlinx.coroutines.CancellationException) {
                // cancel() path — stay inert.
            }
        }
    }

    private fun finish(withError: Boolean) {
        if (finished) return
        finished = true
        reconnectJob?.cancel()
        reconnectJob = null
        collectJob?.cancel()
        collectJob = null
        deps.finalize(withError)
    }
}
